crates/services/importer/src/importer.rs

```diff
@@ -0,0 +1,339 @@
+        let block = &result.sealed_block.entity;
+        let consensus = &result.sealed_block.consensus;
+        let block_id = block.id();
+        let actual_next_height = *block.header().height();
+
+        // During importing of the genesis block, the database should not be initialized
+        // and the genesis block defines the next height.
+        // During the production of the non-genesis block, the next height should be underlying
+        // database height + 1.
+        let expected_next_height = match consensus {
+            Consensus::Genesis(_) => {
+                let result = self.database.latest_block_height();
+                let found = !result.is_not_found();
+                // Because the genesis block is not committed, it should return non found error.
+                // If we find the latest height, something is wrong with the state of the database.
+                if found {
+                    return Err(Error::InvalidUnderlyingDatabaseGenesisState)
+                }
+                actual_next_height
+            }
+            Consensus::PoA(_) => {
+                if actual_next_height == BlockHeight::from(0u32) {
+                    return Err(Error::ZeroNonGenericHeight)
+                }
+
+                let last_db_height = self.database.latest_block_height()?;
+                last_db_height
+                    .checked_add(1u32)
+                    .ok_or(Error::Overflow)?
+                    .into()
+            }
+        };
+
+        if expected_next_height != actual_next_height {
+            return Err(Error::IncorrectBlockHeight(
+                expected_next_height,
+                actual_next_height,
+            ))
+        }
+
+        let db_after_execution = db_tx.as_mut();
+
+        // Importer expects that `UncommittedResult` contains the result of block
+        // execution(It includes the block itself).
+        let actual_height = db_after_execution.latest_block_height()?;
+        if expected_next_height != actual_height {
+            return Err(Error::InvalidDatabaseStateAfterExecution(
+                expected_next_height,
+                actual_height,
+            ))
+        }
+
+        db_after_execution
+            .seal_block(&block_id, &result.sealed_block.consensus)?
+            .should_be_unique(&expected_next_height)?;
+
+        // TODO: This should use a proper BMT MMR. Based on peaks stored in the database,
+        //  we need to calculate a new root. The data type that will do that should live
+        //  in the `fuel-core-storage` or `fuel-merkle` crate.
+        let root = ephemeral_merkle_root(
+            vec![*block.header().prev_root(), block_id.into()].iter(),
+        );
+        db_after_execution
+            .insert_block_header_merkle_root(&expected_next_height, &root)?
+            .should_be_unique(&expected_next_height)?;
+
+        db_tx.commit()?;
+
+        let _ = self.broadcast.send(Arc::new(result));
```

Comment:
Who is waiting on the other end of this channel? If this fails to send is the result lost or can it be retrieved from the db?

Reply:
There is `TxPool` right now, but after maybe it would be one more - GraphQL service.

The `SealedBlock` is available from the database; transaction status is available in the current code, but we will remove it in the future because it's not part of the blockchain. So if the receiver is slow, it could lose some data. It is why I set it for now as `1 << 10`, but maybe we need to increase it in the future or make sure that receivers are not slow=)